IFC_Adapter/CRUD/Read.cs

```diff
@@ -0,0 +1,117 @@
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Adapters.IFC;
+using BH.oM.Adapter;
+using BH.oM.Adapters.IFC;
+using BH.oM.Base;
+using BH.oM.Data.Requests;
+using BH.oM.Geometry;
+using System.Collections.Generic;
+using System.Linq;
+using Xbim.Common.Geometry;
+using Xbim.Ifc2x3.Interfaces;
+using Xbim.ModelGeometry.Scene;
+
+namespace BH.Adapter.IFC
+{
+    public partial class IfcAdapter : BHoMAdapter
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        protected override IEnumerable<IBHoMObject> Read(IRequest request, ActionConfig actionConfig = null)
+        {
+            if (request == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("The request is null, pull failed.");
+                return new List<IBHoMObject>();
+            }
+
+            // Get the toolkit-specific PullConfig
+            IfcPullConfig config = actionConfig as IfcPullConfig;
+            if (config == null)
+            {
+                config = new IfcPullConfig();
+                BH.Engine.Reflection.Compute.RecordNote("Config has not been specified, default config is used.");
+            }
+
+            // Get the settings
+            IfcSettings settings = this.IFCSettings.DefaultIfNull();
+
+            // Get the discipline coming from the request/PullConfig
+            Discipline? requestDiscipline = request.Discipline(config.Discipline);
+            if (requestDiscipline == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Conflicting disciplines have been detected.");
+                return new List<IBHoMObject>();
+            }
+
+            Discipline discipline = requestDiscipline.Value;
```

Comment:
As per our conversation yesterday, worth leaving a comment here.

I feel that this mechanism is currently understandable only after discussion with the author.

We basically have 2 input sources for the `Discipline`:
1. The PullConfig: drives what `Discipline` the objects should be **converted to** by the Adapter
2. The IRequest: drives what `Discipline` to consider as **a source for pulling** the objects

It follows that the `Discipline` input in 1 and 2 may have some degree of incompatibility, hence this required check.

I argued that this feels like an inconsistency/code smell and that I would like to see the `Discipline` input disappear from the Request, to be left to the PullConfig alone. @pawelbaran explained that this would mean less clarity for the users.

Given that this is a workflow that largely replicates the structure of Revit_Toolkit, which has been proved, I will not oppose this approach. However I do believe that there is room for improvement, potentially rethinking the whole mechanism, and that some detailed explanation of the process and reasons behind this would benefit both developers and users - perhaps adding a short wiki page.

Reply:
100% agreed with @alelom, #5 raised as follow-up to this comment.